### PR TITLE
Fix local volumes with common GID, and other fixes

### PR DESCRIPTION
**Pull request by svenssonaxel**
My suggested fixes to https://github.com/logicalclocks/rondb-docker/pull/8

### CODE REVIEW

File: .github/workflows/test_and_deploy.yml [hunk 1]
```diff
@@ -33,10 +33,10 @@ jobs:
             --num-mysql-nodes 1 \
             --num-api-nodes 1 \
             --run-benchmark sysbench_single \
-            --bench-dirs-in-volumes \
+            --volumes-in-local-dir \
```
Comment (olapiv) [inline .github/workflows/test_and_deploy.yml:36]: Bit of a formality, but if the API gets changed, you can run build_run_docker.sh with `--save-sample-files`, to fix the sample files. They're just for links in blog posts and documentation. Or you just revert the changes that I had made in those files.

File: resources/config_templates/my.cnf [hunk 2]
```diff
@@ -81,7 +81,7 @@ ndb-batch-size=1M
 [mysql_cluster]
 # set connectstring to ndb management server (used by all executables)
 ndb-connectstring=%s
-# ndb-wait-setup=10800
+ndb-wait-setup=900
```
Comment (olapiv) [inline resources/config_templates/my.cnf:84]: What's this for?

File: resources/entrypoints/entrypoint.sh [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+echo "[Entrypoint] RonDB Docker Image"
+
+# The host user (which executes build_run_docker.sh) might not have the same UID
+# as mysql inside the container. In order for both users to be able to read and
+# write in the mounted volumes, we add mysql to a group with the same GID as the
+# host user's group.
+if [ "$(getent group "$HOST_GROUP_ID")" ]; then
+	echo "group $HOST_GROUP_ID exists."
+else
+	echo "group $HOST_GROUP_ID does not exist."
```
Comment (olapiv) [inline resources/entrypoints/entrypoint.sh]: Just append `[Entrypoint]` to the echos for consistency.

File: resources/entrypoints/mysqld.sh [hunk 4]
```diff
@@ -148,30 +150,30 @@ if [ "$MYSQL_USER" ]; then
...
     esac
 done
 
 # When using a local socket, mysqladmin shutdown will only complete when the
 # server is actually down.
-echo '[Entrypoint] Shutting down MySQLd via mysqladmin...'
-mysqladmin -uroot --password=$MYSQL_ROOT_PASSWORD shutdown --socket="$SOCKET"
-echo "[Entrypoint] Successfully shut down MySQLd"
+echo '[entrypoints/mysqld.sh] Shutting down MySQLd via mysqladmin...'
+mysqladmin -uroot --password="$MYSQL_ROOT_PASSWORD" shutdown --socket="$SOCKET"
```
Comment (olapiv) [inline resources/entrypoints/mysqld.sh:174]: There a lot of these changes - I guess it's best practice? In case the variable has a whitespace stored?

File: resources/entrypoints/main.sh [hunk 5]
```diff
@@ -40,13 +57,38 @@ else
 
 	set -- "$@" --nodaemon
 	if [ "$1" == "ndb_mgmd" ]; then
-		echo "[Entrypoint] Starting ndb_mgmd"
-		set -- "$@" -f $RONDB_DATA_DIR/config.ini --configdir=$RONDB_DATA_DIR/log
+		echo "[entrypoints/main.sh] Starting ndb_mgmd"
+		set -- "$@" -f "$RONDB_DATA_DIR/config.ini" --configdir="$RONDB_DATA_DIR/log"
```
Comment (olapiv) [inline resources/entrypoints/main.sh:61]: What about the permissions of the logs/data that the mgmd creates?

File: build_run_docker.sh [hunk 6]
```diff
@@ -347,58 +333,43 @@ AUTOBENCH_DBT2_TEMPLATE=$(cat ./resources/config_templates/autobench_dbt2.conf)
...
-
 #######################
 #######################
 #######################
 
 echo "Filling out templates"
 
-source $SCRIPT_DIR/docker.env
-source $SCRIPT_DIR/misc_configs.env
+# shellcheck source=./docker.env
```
Comment (olapiv) [inline build_run_docker.sh:367]: You want to keep this here?
Also, what if the script is executed from somewhere else, does it still work?

File: build_run_docker.sh [hunk 7]
```diff
@@ -278,11 +257,10 @@ RONDB_VERSION_NO_DOT=$(echo "$RONDB_VERSION" | tr -d '.')
 FILE_SUFFIX="v${RONDB_VERSION_NO_DOT}_m${NUM_MGM_NODES}_g${NODE_GROUPS}_r${REPLICATION_FACTOR}_my${NUM_MYSQL_NODES}_api${NUM_API_NODES}"
 
 AUTOGENERATED_FILES_DIR="$SCRIPT_DIR/autogenerated_files/$FILE_SUFFIX"
-rm -rf $AUTOGENERATED_FILES_DIR
```
Comment (olapiv) [inline build_run_docker.sh:281]: Why not just delete it all? Are you expecting to restart a cluster with previous data?

File: build_run_docker.sh [hunk 8]
```diff
@@ -347,58 +333,43 @@ AUTOBENCH_DBT2_TEMPLATE=$(cat ./resources/config_templates/autobench_dbt2.conf)
...
 
-# We add volumes to the data dir for debugging purposes
 VOLUME_TEMPLATE="
-      - %s:%s"
+      - type: %s
+        source: %s
+        target: %s"
 
 ENV_FIELD="
       environment:"
```
Comment (olapiv) [inline build_run_docker.sh]: You could just add `HOST_GROUP_ID` here since all containers should have it.

File: build_run_docker.sh [hunk 9]
```diff
@@ -416,10 +387,10 @@ VOLUMES=()
 add_volume_to_template() {
     local VOLUME_NAME="$1"
     local TARGET_DIR_PATH="$2"
-    local VOLUME_TYPE="$3"
-    local IS_NDBD_DATA_DIR="$4"
-    if [ "$VOLUME_TYPE" == "local" ]; then
+    local IS_NDBD_DATA_DIR="$3"
+    if [ "$VOLUME_TYPE" == local ]; then
         local VOLUME_DIR="$LOCAL_VOLUMES_DIR/$VOLUME_NAME"
+        rm -rf "$VOLUME_DIR"
```
Comment (olapiv) [inline build_run_docker.sh]: Might as well just delete all autogenerated files in the beginning?

**Review by olapiv — CHANGES_REQUESTED** [review]
I'll test it now, but so far some comments..!

File: build_run_docker.sh [hunk 10]
```diff
@@ -746,7 +724,12 @@ else
     exit 1
 fi
 
+# Make directories accessible by group members. This is used as a workaround
+# since the local (host) user likely has a different UID from the user in the
+# docker container.
+chmod g=u -R "$AUTOGENERATED_FILES_DIR"
```
Comment (olapiv) [inline build_run_docker.sh]: This returns `chmod: -R: No such file or directory` on Mac

Comment (olapiv) [reply build_run_docker.sh]:
```bash
chmod --help           
chmod: illegal option -- -
usage:  chmod [-fhv] [-R [-H | -L | -P]] [-a | +a | =a  [i][# [ n]]] mode|entry file ...
        chmod [-fhv] [-R [-H | -L | -P]] [-E | -C | -N | -i | -I] file ...
```

File: resources/entrypoints/main.sh [hunk 11]
```diff
@@ -40,13 +57,38 @@ else
...
+			# modestring.
+			sed -zr '/^(.)\1/d; s/^... //;' |
+			# xargs: Run chmod with an efficient number of file
+			# arguments to correct the group's permissions.
+			xargs -r0 chmod -cf g=u ||
+			# Make sure the process does not exit due to some
+			# failure.
+			true
+			sleep 5
+		done &
```
Comment (olapiv) [inline resources/entrypoints/main.sh:89]: Btw, what happens if the ndbmtd crashes? Could be that the last files are not deletable anymore? Which PID does this even use if the ndbmtd is started with `exec`?

File: .github/workflows/test_and_deploy.yml (hunk 1 above)
Comment (svenssonaxel) [reply .github/workflows/test_and_deploy.yml:36]: Fixed, will push soon

File: resources/config_templates/my.cnf (hunk 2 above)
Comment (svenssonaxel) [reply resources/config_templates/my.cnf:84]: On my machine, sometimes, `mysqld_X` would fail, complaining that data nodes weren't ready in `30s` and suggested increasing this value.

File: resources/entrypoints/entrypoint.sh (hunk 3 above)
Comment (svenssonaxel) [reply resources/entrypoints/entrypoint.sh]: Fixed

File: resources/entrypoints/mysqld.sh (hunk 4 above)
Comment (svenssonaxel) [reply resources/entrypoints/mysqld.sh:174]: Yes, most of these are part of commit `Lint shell scripts`. I used `shellcheck`, a linter for bash and other shells, to find those.

File: resources/entrypoints/main.sh (hunk 5 above)
Comment (svenssonaxel) [reply resources/entrypoints/main.sh:61]: The `umask` value of `0002` gets inherited by all processes started from this script, also `ndb_mgmd`, see PID `31` in this example:
```
docker exec mgmd_1 bash -c 'ps -fuxa; grep -H Umask /proc/*/status'
USER         PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root          58  0.0  0.0   4360  3300 ?        Ss   07:04   0:00 bash -c ps -fuxa; grep -H Umask /proc/*/status
root          64  0.0  0.0   7060  1564 ?        R    07:04   0:00  \_ ps -fuxa
root           1  0.0  0.0   4360  3304 ?        Ss   07:02   0:00 /bin/bash ./docker_entrypoints/rondb_standalone/entrypoint.sh ndb_mgmd --ndb-nodeid=65 --initial
root          30  0.0  0.0   7236  4392 ?        S    07:02   0:00 sudo -E -u mysql /srv/hops/docker_entrypoints/rondb_standalone/main.sh ndb_mgmd --ndb-nodeid=65 --initial
mysql         31  0.8  0.0 1035104 25040 ?       Sl   07:02   0:01  \_ ndb_mgmd --ndb-nodeid=65 --initial --nodaemon -f /srv/hops/mysql-cluster/config.ini --configdir=/srv/hops/mysql-cluster/log
/proc/1/status:Umask:   0022
/proc/30/status:Umask:  0022
/proc/31/status:Umask:  0002
/proc/58/status:Umask:  0022
/proc/self/status:Umask:        0022
/proc/thread-self/status:Umask: 0022
```
Never had a problem with mgmd in testing. It appears that `ndb_mgmd` either does not create any files in the volumes or uses system default permissions, which will result in `0666 & ~UMASK = 0664` for files and `0777 & ~UMASK = 0775` for directories.

File: build_run_docker.sh (hunk 6 above)
Comment (svenssonaxel) [reply build_run_docker.sh:367]: Yes, it's a linter directive, and is needed to not get warnings about missing variable declarations when they are found in the `.env` files. It's inside a comment so won't affect execution.

File: build_run_docker.sh (hunk 7 above)
Comment (svenssonaxel) [reply build_run_docker.sh:281]: This came along by mistake when I reverted some of your changes. I'll put it back in.

File: build_run_docker.sh (hunk 8 above)
Comment (svenssonaxel) [reply build_run_docker.sh]: Done

File: build_run_docker.sh (hunk 9 above)
Comment (svenssonaxel) [reply build_run_docker.sh]: Yup.

File: build_run_docker.sh (hunk 10 above)
Comment (svenssonaxel) [reply build_run_docker.sh]: Fixed.

File: resources/entrypoints/main.sh (hunk 11 above)
Comment (svenssonaxel) [reply resources/entrypoints/main.sh:89]: > Btw, what happens if the ndbmtd crashes? Could be that the last files are not deletable anymore?

Yes, if ndbmtd exits within 5 seconds after creating a new file we're in trouble as you say. I'll fix it, but that means I can't use `exec` for `ndbmtd`.

> Which PID does this even use if the ndbmtd is started with `exec`?

The `&` after `done` creaes a sub shell, so the background bash process becomes a child of `ndbmtd` which looks kinda funny:
```
docker exec ndbd_1 bash -c 'ps -fuxa; grep -H Umask /proc/*/status'
USER         PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root        1477  0.0  0.0   4360  3176 ?        Ss   07:29   0:00 bash -c ps -fuxa; grep -H Umask /proc/*/status
root        1483  0.0  0.0   7060  1584 ?        R    07:29   0:00  \_ ps -fuxa
root           1  0.0  0.0   4360  3300 ?        Ss   07:02   0:00 /bin/bash ./docker_entrypoints/rondb_standalone/entrypoint.sh ndbmtd --ndb-nodeid=1 --initial --ndb-connectstring=mgmd_1:1186
root          30  0.0  0.0   7236  4348 ?        S    07:02   0:00 sudo -E -u mysql /srv/hops/docker_entrypoints/rondb_standalone/main.sh ndbmtd --ndb-nodeid=1 --initial --ndb-connectstring=mgmd_1:1186
mysql         31  0.0  0.0  29636 10216 ?        S    07:02   0:00  \_ ndbmtd --ndb-nodeid=1 --initial --ndb-connectstring=mgmd_1:1186 --nodaemon
mysql         34  0.0  0.0   4360  1776 ?        S    07:02   0:00      \_ /bin/bash /srv/hops/docker_entrypoints/rondb_standalone/main.sh ndbmtd --ndb-nodeid=1 --initial --ndb-connectstring=mgmd_1:1186
mysql       1476  0.0  0.0   2788  1008 ?        S    07:29   0:00      |   \_ sleep 5
mysql         40 29.3  8.0 6409960 5309384 ?     Sl   07:02   8:06      \_ ndbmtd --ndb-nodeid=1 --initial --ndb-connectstring=mgmd_1:1186 --nodaemon
/proc/1/status:Umask:   0022
/proc/1476/status:Umask:        0002
/proc/1477/status:Umask:        0022
/proc/30/status:Umask:  0022
/proc/31/status:Umask:  0002
/proc/34/status:Umask:  0002
/proc/40/status:Umask:  0002
/proc/self/status:Umask:        0022
/proc/thread-self/status:Umask: 0022

```

Comment (svenssonaxel) [reply resources/entrypoints/main.sh:89]: But of course this will also not be the case after the fix, since I no longer use `exec`.

**Comment by svenssonaxel** [comment]
All of the above now pushed.

File: build_run_docker.sh [hunk 12]
```diff
@@ -624,41 +587,40 @@ if [ $NUM_API_NODES -gt 0 ]; then
...
-            template+="$BIND_AUTOBENCH_SYS_MULTI_TEMPLATE"
-            template+="$BIND_AUTOBENCH_DBT2_SINGLE_TEMPLATE"
-            template+="$BIND_AUTOBENCH_DBT2_MULTI_TEMPLATE"
+        if [ "$VOLUME_TYPE" == "docker" ]; then
+            add_file_to_template "$DBT2_CONF_SINGLE_FILEPATH" "$BENCH_DIR/dbt2_single/dbt2_run_1.conf"
+            add_file_to_template "$DBT2_CONF_MULTI_FILEPATH" "$BENCH_DIR/dbt2_multi/dbt2_run_1.conf"
+            add_file_to_template "$AUTOBENCH_SYS_SINGLE_FILEPATH" "$BENCH_DIR/sysbench_single/autobench.conf"
+            add_file_to_template "$AUTOBENCH_SYS_MULTI_FILEPATH" "$BENCH_DIR/sysbench_multi/autobench.conf"
+            add_file_to_template "$AUTOBENCH_DBT2_SINGLE_FILEPATH" "$BENCH_DIR/dbt2_single/autobench.conf"
+            add_file_to_template "$AUTOBENCH_DBT2_MULTI_FILEPATH" "$BENCH_DIR/dbt2_multi/autobench.conf"
```
Comment (olapiv) [inline build_run_docker.sh]: This fails unless you use the `[ "$NUM_MYSQL_NODES" -gt 1 ]` check, such as this:

```
if [ "$VOLUME_TYPE" == "docker" ]; then
    add_file_to_template "$DBT2_CONF_SINGLE_FILEPATH" "$BENCH_DIR/dbt2_single/dbt2_run_1.conf"
    add_file_to_template "$AUTOBENCH_DBT2_SINGLE_FILEPATH" "$BENCH_DIR/dbt2_single/autobench.conf"
    add_file_to_template "$AUTOBENCH_SYS_SINGLE_FILEPATH" "$BENCH_DIR/sysbench_single/autobench.conf"
    if [ "$NUM_MYSQL_NODES" -gt 1 ]; then
        add_file_to_template "$DBT2_CONF_MULTI_FILEPATH" "$BENCH_DIR/dbt2_multi/dbt2_run_1.conf"
        add_file_to_template "$AUTOBENCH_DBT2_MULTI_FILEPATH" "$BENCH_DIR/dbt2_multi/autobench.conf"
        add_file_to_template "$AUTOBENCH_SYS_MULTI_FILEPATH" "$BENCH_DIR/sysbench_multi/autobench.conf"
    fi
fi
```

**Comment by olapiv** [comment]
I just realised this also changes the API of the script - now, the absence of `-lv` affects all directories, also the benchmarking ones. There are parts in the README which assume that benchmarking directories are always mounted locally. For example:

> The results of the benchmarks are mounted into the local filesystem into the "autogenerated_files" directory. Look for "final_result.txt" in the directory of the benchmark that was run to see the results. For more information on how to read the benchmarking output, refer to the [RonDB documentation](http://docs.rondb.com) once again.

File: resources/entrypoints/main.sh [hunk 13]
```diff
@@ -82,11 +82,23 @@ else
 			# Make sure the process does not exit due to some
 			# failure.
 			true
+		}
+		while true; do
+			ensure-group-permissions
 			sleep 5
 		done &
```
Comment (olapiv) [inline resources/entrypoints/main.sh:89]: When running this, my terminal looks as follows:

```
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/1/T15F1.ctl' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/1/T15F0.ctl' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/1/T17F2.ctl' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/10/T17F1.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/10/T17F0.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/10/T17F3.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/10/T17F2.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/8/T17F1.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/8/T17F0.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/8/T17F3.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/8/T17F2.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/15/T17F1.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/15/T17F0.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/15/T17F3.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/15/T17F2.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/12/T17F1.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/12/T17F0.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/12/T17F3.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/12/T17F2.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/13/T17F1.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/13/T17F0.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/13/T17F3.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/13/T17F2.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/14/T17F1.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/14/T17F0.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/14/T17F3.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/14/T17F2.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
```

Basically I can barely read anything else anymore. Is there any way of hiding this?

Comment (olapiv) [reply resources/entrypoints/main.sh:89]: This happens a fair bit when benchmarks are run. And it is even run when omitting `-lv` (I don't mind that, I just mind the printing).

Comment (olapiv) [reply resources/entrypoints/main.sh:89]: If possible maybe only print stderr

Comment (svenssonaxel) [reply resources/entrypoints/main.sh:89]: I activated printing on purpose, but you're right that it's better not to print. I'll change it.

**Comment by svenssonaxel** [comment]
> I just realised this also changes the API of the script - now, the absence of `-lv` affects all directories, also the benchmarking ones. There are parts in the README which assume that benchmarking directories are always mounted locally.

Which route do you want to take here? I see several possibilities:
* Fix readme to explain `-lv` is needed to see benchmark results
* Change behavior so that benchmark dirs will always mount locally while the flag determines how to handle other volumes
* Remove the possibility of using docker volumes, and always use local mounts for everything.

I have a weak preference for the latter, but that's mostly a matter of taste. I feel like docker volumes are "out of control", or "cannot be owned", a little bit like global variables.

**Comment by olapiv** [comment]
I think you'd also appreciate Docker volumes a bit more if you used Docker Desktop :) It's basically occupied storage that is difficult to forget, since you see it floating around there all the time.

So I'd just change the Readme.

File: build_run_docker.sh (hunk 12 above)
Comment (svenssonaxel) [reply build_run_docker.sh]: Fixed, except I also added `if [ "$NUM_MYSQL_NODES" -gt 0 ]`

File: resources/entrypoints/main.sh (hunk 13 above)
Comment (svenssonaxel) [reply resources/entrypoints/main.sh:89]: Fixed.

**Comment by svenssonaxel** [comment]
> change the Readme.

Fixed.
